src/dbnode/storage/series/reader.go

```diff
@@ -126,39 +138,45 @@ func (r Reader) readersWithBlocksMapAndBuffer(
 					return nil, err
 				}
 				if streamedBlock.IsNotEmpty() {
-					results = append(results, []xio.BlockReader{streamedBlock})
+					resultsBlock = append(resultsBlock, streamedBlock)
 					// NB(r): Mark this block as read now
 					block.SetLastReadTime(now)
 					if r.onRead != nil {
 						r.onRead.OnReadBlock(block)
 					}
 				}
-				continue
+				retrievedFromCache = true
 			}
 		}
 
-		switch {
-		case cachePolicy == CacheAll:
-			// No-op, block metadata should have been in-memory
-		case r.retriever != nil:
-			// Try to stream from disk
-			if r.retriever.IsBlockRetrievable(blockAt) {
-				streamedBlock, err := r.retriever.Stream(ctx, r.id, blockAt, r.onRetrieve)
-				if err != nil {
-					return nil, err
-				}
-				if streamedBlock.IsNotEmpty() {
-					results = append(results, []xio.BlockReader{streamedBlock})
+		if !retrievedFromCache {
+			switch {
+			case cachePolicy == CacheAll:
+				// No-op, block metadata should have been in-memory
+			case r.retriever != nil:
+				// Try to stream from disk
+				if r.retriever.IsBlockRetrievable(blockAt) {
+					streamedBlock, err := r.retriever.Stream(ctx, r.id, blockAt, r.onRetrieve)
+					if err != nil {
+						return nil, err
+					}
+					if streamedBlock.IsNotEmpty() {
+						resultsBlock = append(resultsBlock, streamedBlock)
+					}
 				}
 			}
 		}
-	}
 
-	if seriesBuffer != nil {
-		bufferResults := seriesBuffer.ReadEncoded(ctx, start, end)
-		if len(bufferResults) > 0 {
-			results = append(results, bufferResults...)
+		if seriesBuffer != nil {
+			bufferResults := seriesBuffer.ReadEncoded(ctx, blockAt, blockAt.Add(size))
+			// Multiple block results may be returned here (for the same block
+			// start) - one for warm writes and another for cold writes.
+			for _, bufferRes := range bufferResults {
+				resultsBlock = append(resultsBlock, bufferRes...)
+			}
 		}
+
+		results = append(results, resultsBlock)
```

Comment:
Perhaps only do this if `len(resultsBlock) > 0`?  Otherwise we'll have a lot of nil's for blocks that don't have any results, and each one of them gets sent over RPC.

Reply:
Sure.